### ISSUE TITLE
Spikes in altitude when "MSL instead of WGS84" is enabled

### ISSUE THREAD

**Issue by icrf2000** [mendhak/gpslogger#346]
With version 66, I get random spikes on the altitude when I enable the "MSL instead of WGS84" option. This did not happen with version 63. I recorded a trip today with 66 and 63 running on the same device (Admittedly, the old 63 was built myself with modified Java package name, so I could install it in parallel to the new 66).

I attach a screenshot which illustrates the spikes. I could also provide extracts of the two GPX files, but don't see how to upload them to Github.

![screenshot - 08232015 - 04 53 51 pm](https://cloud.githubusercontent.com/assets/11777822/9428703/91559078-49b7-11e5-8df6-e8fff8c55af0.png)

**Comment by icrf2000** [comment]
Looks like from time to time the GEOIDHEIGHT element is empty and therefore no geoid correction can be done. The mysterious question is why does version 63 running at the same time on the same device not experience this effect.

![screenshot - 08232015 - 08 06 51 pm](https://cloud.githubusercontent.com/assets/11777822/9429509/924142b4-49d2-11e5-8761-f45bda7904d1.png)

**Comment by icrf2000** [comment]
Did a lengthy trip with two devices (Galaxy S4 + Nexus6) both running v63 and v66 in parallel. After this, I must correct my statement that it does not happen with v63. There are such spikes too with this older version. Not at exactly the same places and apparently less frequently, but this might be explained by pure luck - when exactly the app reads data from the sensor. 

I will attempt to close this issue and spare Mendhak from spending time on it. Best solution (at least for me) is probably not to use the "MSL instead of WGS84" option:-(

**Comment by mendhak** [comment]
I've modified it so that is that MSL preference is enabled and a geoidheight is not present, then elevation is not recorded.  This should remove spikes and in its place elevation will be absent.  

I have created an APK if you want to test here: https://www.dropbox.com/s/2u3nipz0zm1xjjs/gpslogger-66-issue346.apk?dl=0

**Comment by icrf2000** [comment]
I've downloaded the APK and will test it tomorrow. At least the idea seems to be very good.

What I already did now: Edit an existing old GPX in a text editor by removing <ELE> for some random points. Postprocessing software (gpsbabel, GoogleEarth) doesn't seem to have problems with such files, so I'm confident that your workaround will do well. Will post here as soon as I tested with a real new track.

(BTW it is still strange that in my older GPX's I find no single TRKPT without GEOIDHEIGHT. But be it... Maybe an Android update... And I didn't look at all old files, so maybe pure chance that I didn't check "broken" ones).

**Comment by mendhak** [comment]
Could be updates, essentially the performance is down to firmware and drivers for those GPS chips.  If they do minor things like performance improvements, tweaks to how modules communicate, or managing how hardware sleeps and wakes up, those things in turn affects how the hardware behaves when queried, which in turn affects accuracy and performance, which in turn affects the GPS positions and associated data.  These problems could even go away and come back, many permutations.

**Comment by icrf2000** [comment]
I recorded a test track: 65km, 923 points in total, 195 without GEOIDHEIGHT, 195 without ELE. Looks good with gpsbabel and GoogleEarth. So I would say your workaround works as great.

I find it alarming that such a high proportion (21%) of points is delivered without GEOIDHEIGHT (and also without HDOP, VDOP, PDOP, but those are not so interesting). But I'm afraid that's specific to my device:-(

**Comment by icrf2000** [comment]
Found it! It was the location sources... I ran the app with the default settings to allow not only GPS, but all location sources (Network/Passive/GPS). But only GPS seems to provide GEOIDHEIGHT. So after I disabled Network and Passive, I have no points without correct elevation anymore, v66 behaves as my old v63 did in the past. Obviously I ran v63 with GPS only and lost this configuration somewhere in the upgrade path.

Why did I not notice this when looking at the GPX......? I could have... But also the `<trkpt>`s without `<geoidheight>`/`<ele>` (I assume now that those came from a non-GPS provider) show `<src>gps</src>` as indicated in the screenshot above. Looking at the code, this "gps" string seems to come directly from the framework

`loc.getProvider()`

so the app is not to blame.

Anyway, your workaround of not showing `<ele>` if the data is incorrect is great and I would suggest to include it in the main branch.
